blockchain/state_manager/src/lib.rs

```diff
@@ -185,9 +241,115 @@ where
+    {
+        span!("state_call_raw", {
+            let block_store = self.get_block_store_ref();
+            let buf_store = BufferedBlockStore::new(block_store);
+            let mut vm = VM::new(
+                bstate,
+                &buf_store,
+                *bheight,
+                DefaultSyscalls::new(&buf_store),
+                rand,
+            )?;
+
+            if msg.gas_limit() == 0 {
+                msg.set_gas_limit(10000000000)
+            }
+
+            let actor = self
+                .get_actor(msg.from(), bstate)?
+                .ok_or_else(|| Error::Other("Could not get actor".to_string()))?;
+            msg.set_sequence(actor.sequence);
+            let apply_ret = vm.apply_implicit_message(msg);
+            trace!(
+                "gas limit {:},gas price {:?},value {:?}",
+                msg.gas_limit(),
+                msg.gas_price(),
+                msg.value()
+            );
+            if let Some(err) = apply_ret.act_error() {
+                warn!("chain call failed: {:?}", err);
+            }
+
+            Ok(InvocResult {
+                msg: msg.clone(),
+                msg_rct: Some(apply_ret.msg_receipt().clone()),
+                actor_error: apply_ret.act_error().map(|e| e.to_string()),
+            })
+        })
+    }
+
+    /// runs the given message and returns its result without any persisted changes.
+    pub fn call(
+        &self,
+        message: &mut UnsignedMessage,
+        tipset: Option<Tipset>,
+    ) -> StateCallResult<UnsignedMessage>
+    where
+        DB: BlockStore,
+    {
+        let ts = if let Some(t_set) = tipset {
+            t_set
+        } else {
+            chain::get_heaviest_tipset(self.get_block_store_ref())
+                .map_err(|_| Error::Other("Could not get heaviest tipset".to_string()))?
+                .ok_or_else(|| Error::Other("Empty Tipset given".to_string()))?
+        };
+        let state = ts.parent_state();
+        let chain_rand = ChainRand::new(ts.key().to_owned());
+        self.call_raw(message, state, &chain_rand, &ts.epoch())
+    }
+
+    /// returns the result of executing the indicated message, assuming it was executed in the indicated tipset.
+    pub fn replay(
+        &self,
+        ts: &Tipset,
+        mcid: &Cid,
+    ) -> Result<(UnsignedMessage, Option<ApplyRet>), Error>
+    where
+        DB: BlockStore,
+    {
+        let mut outm: Option<UnsignedMessage> = None;
+        let mut outr: Option<ApplyRet> = None;
+        let callback = |cid: Cid, unsigned: UnsignedMessage, apply_ret: ApplyRet| {
+            if cid == mcid.clone() {
+                outm = Some(unsigned);
+                outr = Some(apply_ret);
+                return Err("halt".to_string());
+            }
+
+            Ok(())
+        };
+        let result = self.compute_tipset_state(ts.blocks(), Some(callback));
+
+        if let Err(error_message) = result {
+            if error_message.to_string() == "halt" {
```

Comment:
I would prefer not to downcast if possible. It looks like the callback is just using this for halting purposes during execution. I think it is weird as it is because lotus is essentially using an error type as it's halting mechanic. Would have been better imo if a partial application of the sorts was returned. I think it works for now because the checking is only happening when the halt mechanic is applied